engine/src/main/java/org/pentaho/di/trans/steps/terafast/TeraFast.java

```diff
@@ -0,0 +1,475 @@
+    } finally {
+      IOUtils.closeQuietly( control );
+      IOUtils.closeQuietly( this.fastload );
+    }
+  }
+
+  /**
+   * Invoke loading with loading commands.
+   *
+   * @throws KettleException
+   *           ...
+   */
+  private void invokeLoadingCommand() throws KettleException {
+    final FastloadControlBuilder builder = new FastloadControlBuilder();
+    builder.setSessions( this.meta.getSessions().getValue() );
+    builder.setErrorLimit( this.meta.getErrorLimit().getValue() );
+    builder.logon( this.meta.getDbMeta().getHostname(), this.meta.getDbMeta().getUsername(), this.meta
+      .getDbMeta().getPassword() );
+    builder.setRecordFormat( FastloadControlBuilder.RECORD_VARTEXT );
+    try {
+      builder.define(
+        this.meta.getRequiredFields( this.getTransMeta() ), meta.getTableFieldList(), resolveFileName( this.meta
+          .getDataFile().getValue() ) );
+    } catch ( Exception ex ) {
+      throw new KettleException( "Error defining data file!", ex );
+    }
+    builder.show();
+    builder.beginLoading( this.meta.getDbMeta().getPreferredSchemaName(), this.meta.getTargetTable().getValue() );
+
+    builder.insert( this.meta.getRequiredFields( this.getTransMeta() ), meta.getTableFieldList(), this.meta
+      .getTargetTable().getValue() );
+    builder.endLoading();
+    builder.logoff();
+    final String control = builder.toString();
+    try {
+      logDetailed( "Control file: " + control );
+      IOUtils.write( control, this.fastload );
+      // this.fastload.flush();
+    } catch ( IOException e ) {
+      throw new KettleException( "Error while execution control command [controlCommand=" + control + "]", e );
+    } finally {
+      IOUtils.closeQuietly( this.fastload );
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.pentaho.di.trans.step.BaseStep#dispose(org.pentaho.di.trans.step.StepMetaInterface,
+   *      org.pentaho.di.trans.step.StepDataInterface)
+   */
+  @Override
+  public void dispose( final StepMetaInterface smi, final StepDataInterface sdi ) {
+    this.meta = (TeraFastMeta) smi;
+    // this.data = (GenericStepData) sdi;
+
+    try {
+      if ( this.fastload != null ) {
+        IOUtils.write( new FastloadControlBuilder().endLoading().toString(), this.fastload );
+      }
+    } catch ( IOException e ) {
+      logError( "Unexpected error encountered while issuing END LOADING", e );
+    }
+    IOUtils.closeQuietly( this.dataFile );
+    IOUtils.closeQuietly( this.fastload );
+    try {
+      if ( this.process != null ) {
+        int exitValue = this.process.waitFor();
+        logDetailed( "Exit value for the fastload process was : " + exitValue );
+        if ( exitValue != 0 ) {
+          logError( "Exit value for the fastload process was : " + exitValue );
+          setErrors( DEFAULT_ERROR_CODE );
+        }
+      }
+    } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)